Book_apis/models.py

```diff
@@ -1,3 +1,95 @@
+from django.contrib.auth.models import AbstractBaseUser, PermissionsMixin
 from django.db import models
 
-# Create your models here.
+from .managers import UserManager
+
+
+class User(AbstractBaseUser, PermissionsMixin):
+
+    USER_ROLES = [
+        ('ADMIN', 'Admin'),
+        ('CUSTOMER', 'Customer'),
+        ('VENDOR', 'Vendor')
+    ]
+    created_at = models.DateTimeField(auto_now_add=True)
+    date_of_birth = models.DateField(null=True)
+    email = models.EmailField(max_length=255, unique=True)
+    first_name = models.CharField(max_length=350)
+    last_name = models.CharField(max_length=350)
+    phone = models.CharField(max_length=350, null=True, unique=True)
+    username = models.CharField(max_length=350, null=True)
+    user_role = models.CharField(
+        max_length=20,
+        choices=USER_ROLES,
+        default='CUSTOMER'
+    )
+    is_active = models.BooleanField(default=True)
+    is_staff = models.BooleanField(default=False)
+
+    objects = UserManager()
+
+    USERNAME_FIELD = 'email'
+    REQUIRED_FIELDS = ['first_name', 'last_name']
+
+    def get_full_name(self):
+        return f'{self.first_name} {self.last_name}'
+    
+    def __str__(self):
+        return self.email
+    
+
+class Publisher(models.Model):
+    name = models.CharField(max_length=100)
+    address = models.TextField()
+
+    def __str__(self):
+        return self.name
+
+
+class Shop(models.Model):
+    name = models.CharField(max_length=100)
+    address = models.TextField()
+    
+    def __str__(self):
+        return self.name   
+    
+
+class Book(models.Model):
+    
+    title = models.CharField(max_length=100)
+    isbn = models.CharField(max_length=250, unique=True)
+    description = models.CharField(max_length=500)
+    publish_date = models.DateField(auto_now_add=True)
+    is_available = models.BooleanField(default=True)
+    price=models.IntegerField()
+    author = models.CharField(max_length=250)
+    publisher = models.ForeignKey(Publisher, on_delete=models.CASCADE)
+    shop = models.ForeignKey(Shop, on_delete=models.CASCADE)
+    
+    def __str__(self):
+        return self.title
+    
+
+class Cart(models.Model):
+    user = models.ForeignKey(User, on_delete=models.CASCADE)
+    books = models.ManyToManyField(Book)
```

Comment:
I think many to many fields will cause the extra cost for checking the stocks and other validations. I think it will be better if we create another model associated with a cart for storing the books.